sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -169,11 +169,21 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   private def insertInto(tableIdent: TableIdentifier): Unit = {
     val partitions = partitioningColumns.map(_.map(col => col -> (None: Option[String])).toMap)
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation schema's can be different from the input logicalPlan, since
+    // partition columns are all moved after data column. We Project to adjust the ordering.
+    // TODO: this belongs in the analyzer.
```

Comment:
Nit: some typos

- "... relation schema's ..." => "... relation's schema ..."
- "... after data column." => "... after data columns"
- "... belongs in ..." => "... belongs to ..."